src/shared/util.js

```diff
@@ -1218,20 +1218,164 @@ function MessageHandler(sourceName, targetName, comObj) {
   this.sourceName = sourceName;
   this.targetName = targetName;
   this.comObj = comObj;
-  this.callbackIndex = 1;
+  this.callbackId = 1;
+  this.streamId = 1;
   this.postMessageTransfers = true;
-  var callbacksCapabilities = this.callbacksCapabilities = Object.create(null);
-  var ah = this.actionHandler = Object.create(null);
+  let streamSinks = this.streamSinks = Object.create(null);
+  let streamControllers = this.streamControllers = Object.create(null);
+  let callbacksCapabilities = this.callbacksCapabilities = Object.create(null);
+  let ah = this.actionHandler = Object.create(null);
+
+  let _processStreamMessage = (data) => {
+    let sourceName = this.sourceName;
+    let targetName = data.sourceName;
+    let streamId = data.streamId;
+    let sendStreamResponse = ({ stream, success, reason }) => {
+      comObj.postMessage({ sourceName, targetName, stream,
+                           success, streamId });
+    };
+    let deleteStreamController = () => {
+      // Delete streamController only when start, pull and
+      // cancel callbacks are resolved, to avoid "TypeError".
+      Promise.all([
+        streamControllers[data.streamId].startCall,
+        streamControllers[data.streamId].pullCall,
+        streamControllers[data.streamId].cancelCall
+      ]).then(() => {
+        delete streamControllers[data.streamId];
+      });
+    };
+    switch (data.stream) {
+      case 'start_complete':
+        if (data.success) {
+          streamControllers[data.streamId].startCall.resolve();
+        } else {
+          streamControllers[data.streamId].startCall.reject();
+        }
+        break;
+      case 'pull_complete':
+        if (data.success) {
+          streamControllers[data.streamId].pullCall.resolve();
+        } else {
+          streamControllers[data.streamId].pullCall.reject();
+        }
+        break;
+      case 'pull':
+        // Ignore any pull after close is called.
+        if (!streamSinks[data.streamId]) {
+          sendStreamResponse({ stream: 'pull_complete', success: true });
+          break;
+        }
+        // Pull increases the desiredSize property of sink,
+        // so when it changes from negative to positive,
+        // set ready property as resolved promise.
+        if (streamSinks[data.streamId].desiredSize <= 0 &&
+            data.desiredSize > 0) {
+          streamSinks[data.streamId].sinkCapability.resolve();
+        }
+        // Reset desiredSize property of sink on every pull.
+        streamSinks[data.streamId].desiredSize = data.desiredSize;
+        let pullResult = streamSinks[data.streamId].onPull !== null ?
+          streamSinks[data.streamId].onPull() : undefined;
+        Promise.resolve(pullResult).then(() => {
```

Comment:
add `let resolveCall = (fn, args) => { if (!fn) return Promise.resolve(undefined); return new Promise( (resolve, reject) => { resolve(fn.apply(null, args)); }); }` and use it in pull, cancel and start cases.